File: 1682_longest_palindromic_subsequence_ii/solution.cpp
// LeetCode 1682 - Longest Palindromic Subsequence II
// https://leetcode.com/problems/longest-palindromic-subsequence-ii/

#include <algorithm>
#include <string>
#include <vector>

class Solution {
public:
    int longestPalindromeSubseq(std::string s) {
        int n = static_cast<int>(s.size());
        std::vector<std::vector<std::vector<int>>> dp(
            n, std::vector<std::vector<int>>(n, std::vector<int>(26, 0)));
        for (int length = 2; length <= n; ++length) {
            for (int i = 0; i + length - 1 < n; ++i) {
                int j = i + length - 1;
                for (int c = 0; c < 26; ++c) {
                    dp[i][j][c] = std::max(dp[i + 1][j][c], dp[i][j - 1][c]);
                }
                if (s[i] == s[j]) {
                    int c = s[i] - 'a';
                    int inner = 0;
                    if (length > 2) {
                        for (int x = 0; x < 26; ++x) {
                            if (x != c) {
                                inner = std::max(inner, dp[i + 1][j - 1][x]);
                            }
                        }
                    }
                    dp[i][j][c] = std::max(dp[i][j][c], inner + 2);
                }
            }
        }
        return *std::max_element(dp[0][n - 1].begin(), dp[0][n - 1].end());
    }
};

File: 1682_longest_palindromic_subsequence_ii/Solution.scala
// LeetCode 1682 - Longest Palindromic Subsequence II
// https://leetcode.com/problems/longest-palindromic-subsequence-ii/

object Solution {
  def longestPalindromeSubseq(s: String): Int = {
    val n = s.length
    val dp = Array.ofDim[Int](n, n, 26)
    for (length <- 2 to n; i <- 0 to n - length) {
      val j = i + length - 1
      for (c <- 0 until 26) {
        dp(i)(j)(c) = math.max(dp(i + 1)(j)(c), dp(i)(j - 1)(c))
      }
      if (s(i) == s(j)) {
        val c = s(i) - 'a'
        var inner = 0
        if (length > 2) {
          for (x <- 0 until 26 if x != c) {
            if (dp(i + 1)(j - 1)(x) > inner) inner = dp(i + 1)(j - 1)(x)
          }
        }
        if (inner + 2 > dp(i)(j)(c)) dp(i)(j)(c) = inner + 2
      }
    }
    dp(0)(n - 1).max
  }
}
